src/framework/ui/qml/Muse/Ui/initialletternavigation.cpp

```diff
@@ -0,0 +1,202 @@
+{
+    return m_panel;
+}
+
+void InitialLetterNavigation::setPanel(NavigationPanel* panel)
+{
+    if (panel == m_panel) {
+        return;
+    }
+    m_panel = panel;
+    emit panelChanged();
+}
+
+int InitialLetterNavigation::controlColumn() const
+{
+    return m_controlColumn;
+}
+
+void InitialLetterNavigation::setControlColumn(int controlColumn)
+{
+    if (controlColumn == m_controlColumn) {
+        return;
+    }
+    m_controlColumn = controlColumn;
+    emit controlColumnChanged();
+}
+
+bool InitialLetterNavigation::eventFilter(QObject* watched, QEvent* event)
+{
+    if (!event || event->type() != QEvent::ShortcutOverride) {
+        return QObject::eventFilter(watched, event);
+    }
+
+    const QKeyEvent* keyEvent = static_cast<QKeyEvent*>(event);
+
+    const bool eventKeyIsSpace = keyEvent->key() == Qt::Key::Key_Space;
+    const bool isNavKey = NAVIGATION_KEYS.contains((Qt::Key)keyEvent->key());
+    if (isNavKey || (eventKeyIsSpace && m_inputString.isEmpty())) {
+        // Ignore all navigation keys apart from space (unless space is the first thing we type)...
+        clearBuffer();
+        return QObject::eventFilter(watched, event);
+    }
+
+    const QString eventText = normalizeForSearch(keyEvent->text());
+
+    static const QRegularExpression controlCharsExceptSpace("[\\x00-\\x1F\\x7F]");
+    if (eventText.isEmpty() || eventText.contains(controlCharsExceptSpace)) {
+        // Ignore text containing ASCII control characters...
+        clearBuffer();
+        return QObject::eventFilter(watched, event);
+    }
+
+    event->accept();
+    m_inputBufferTimer.start();
+    m_inputString.append(eventText);
+
+    // If the entire string consists of the same char (e.g. "eeee"), we "cycle" through
+    // entries starting with that char using m_nextStartIndex...
+    const bool cycling = isRepetition(m_inputString);
+    const int indexOfMatch = indexForInput(eventText, cycling);
+
+    if (indexOfMatch < 0) {
+        clearBuffer();
+        return !eventKeyIsSpace; // Ignore space if it's the last character typed and there are no matches...
+    }
+
+    if (cycling) {
+        m_nextStartIndex = indexOfMatch + 1;
+    }
+
+    //! NOTE: This has to happen before navigateToIndex in some cases - the view might be
+    //! deleting navigation controls when they're not visible...
+    emit requestVisible(indexOfMatch);
+
+    const INavigation::Index navIndex = { m_controlColumn, indexOfMatch };
+    navigateToIndex(navIndex);
+
+    return true;
+}
+
+int InitialLetterNavigation::indexForInput(const QString& eventText, bool cycling) const
+{
+    for (int i = m_nextStartIndex; i < m_stringList.size(); ++i) {
+        const QString& itemString = normalizeForSearch(m_stringList.at(i));
+        const QString& searchString = cycling ? eventText : m_inputString;
+        if (itemString.startsWith(searchString)) {
+            return i;
+        }
+    }
+    return -1;
+}
```

Comment:
I don't think we should store `m_nextStartIndex`. Instead we should use the dropdown's current index, and cache it in this class as `m_currentIndex` if necessary. Then the logic would become...

```C++
int InitialLetterNavigation::indexForInput(const QString& eventText, bool cycling) const
{
    const int startIndex = m_currentIndex + (cycling ? 1 : 0);

    for (int i = startIndex; i < m_stringList.size(); ++i) {
        // return if match found
    }

    for (int i = 0; i < startIndex; ++i) {
        // return if match found
    }

    return -1;
}
```

We need to loop twice in order to check from the beginning if there's no match after the current item.